packages/twenty-server/src/modules/mail-sender/workflow-actions/send-email.workflow-action.ts

```
@@ -4,22 +4,64 @@ import { z } from 'zod';
 import { EmailService } from 'src/engine/core-modules/email/email.service';
+import { InjectObjectMetadataRepository } from 'src/engine/object-metadata-repository/object-metadata-repository.decorator';
+import { ConnectedAccountWorkspaceEntity } from 'src/modules/connected-account/standard-objects/connected-account.workspace-entity';
+import { ConnectedAccountRepository } from 'src/modules/connected-account/repositories/connected-account.repository';
+import {
+  WorkflowStepExecutorException,
+  WorkflowStepExecutorExceptionCode,
+} from 'src/modules/workflow/workflow-executor/exceptions/workflow-step-executor.exception';
+import { ScopedWorkspaceContextFactory } from 'src/engine/twenty-orm/factories/scoped-workspace-context.factory';
+import {
+  MailSenderException,
+  MailSenderExceptionCode,
+} from 'src/modules/mail-sender/exceptions/mail-sender.exception';
+import { GmailClientProvider } from 'src/modules/messaging/message-import-manager/drivers/gmail/providers/gmail-client.provider';
 
 @Injectable()
 export class SendEmailWorkflowAction {
   private readonly logger = new Logger(SendEmailWorkflowAction.name);
   constructor(
     private readonly environmentService: EnvironmentService,
     private readonly emailService: EmailService,
+    private readonly gmailClientProvider: GmailClientProvider,
+    private readonly scopedWorkspaceContextFactory: ScopedWorkspaceContextFactory,
+    @InjectObjectMetadataRepository(ConnectedAccountWorkspaceEntity)
+    private readonly connectedAccountRepository: ConnectedAccountRepository,
   ) {}
 
+  private async getEmailClient(step: WorkflowSendEmailStep) {
+    const { workspaceId } = this.scopedWorkspaceContextFactory.create();
+
+    if (!workspaceId) {
+      throw new WorkflowStepExecutorException(
+        'Scoped workspace not found',
+        WorkflowStepExecutorExceptionCode.SCOPED_WORKSPACE_NOT_FOUND,
+      );
+    }
+
+    const connectedAccount =
+      await this.connectedAccountRepository.getByIdOrFail(
```

let's throw a specific exception there